### PR TITLE
feat(rfq-relayer): add MaxBalance param

### PR DESCRIPTION
**Pull request by dwasse**
## Summary by CodeRabbit

- **New Features**
	- Introduced minimum and maximum balance configuration for token quoting, allowing for more precise management of available funds.
	- Added functionality to retrieve minimum and maximum balances for specific tokens based on their configuration.
  
- **Bug Fixes**
	- Enhanced quoting logic to ensure that minimum and maximum balances are preserved when calculating quote amounts.
  
- **Tests**
	- Updated test cases to account for new minimum and maximum balance parameters, ensuring accurate validation of quoting behavior.

### CODE REVIEW

File: services/rfq/relayer/quoter/quoter.go [hunk 1]
```diff
@@ -519,13 +519,17 @@ func (m *Manager) getOriginAmount(parentCtx context.Context, origin, dest int, a
 		quoteAmount = minQuoteAmount
 	}
 
-	// Finally, clip the quoteAmount by the balance
-	if quoteAmount.Cmp(balance) > 0 {
-		span.AddEvent("quote amount greater than balance", trace.WithAttributes(
+	// Finally, clip the quoteAmount by the minimum balance
+	minBalance := m.config.GetMinBalance(dest, address)
+	quotableBalance := new(big.Int).Sub(balance, minBalance)
```
Comment (ChiTimesChi) [inline services/rfq/relayer/quoter/quoter.go:524]: Hm, what if our balance is lower than minBalance?

File: services/rfq/relayer/quoter/quoter.go [hunk 2]
```diff
@@ -519,17 +550,32 @@ func (m *Manager) getOriginAmount(parentCtx context.Context, origin, dest int, a
 		quoteAmount = minQuoteAmount
 	}
 
-	// Finally, clip the quoteAmount by the balance
-	if quoteAmount.Cmp(balance) > 0 {
-		span.AddEvent("quote amount greater than balance", trace.WithAttributes(
+	// Clip the quoteAmount by the max origin balance
+	maxBalance := m.config.GetMaxBalance(input.OriginChainID, input.OriginTokenAddr)
+	if input.OriginBalance != nil && maxBalance.Cmp(big.NewInt(0)) > 0 {
+		quotableBalance := new(big.Int).Sub(maxBalance, input.OriginBalance)
```
Comment (ChiTimesChi) [inline services/rfq/relayer/quoter/quoter.go:556]: I think we need to check if `quotableBalance < 0` here, as we could assign a negative value to `quoteAmount` below should the balance exceed maximum.

File: services/rfq/relayer/relconfig/getters.go [hunk 3]
```diff
@@ -298,6 +298,43 @@ func (c Config) GetQuoteOffsetBps(chainID int, tokenName string, isOrigin bool)
 	return offset, nil
 }
 
+const defaultMaxBalance = 0
```
Comment (ChiTimesChi) [inline services/rfq/relayer/relconfig/getters.go:301]: Should we use something that resembles "positive infinity" here instead as a default value? This way we could actually use maxBalance = 0 to effectively stop quoting "to chain", while still quoting the other direction if that makes sense.

File: services/rfq/relayer/quoter/quoter.go [hunk 4]
```diff
@@ -519,17 +550,32 @@ func (m *Manager) getOriginAmount(parentCtx context.Context, origin, dest int, a
 		quoteAmount = minQuoteAmount
 	}
 
-	// Finally, clip the quoteAmount by the balance
-	if quoteAmount.Cmp(balance) > 0 {
-		span.AddEvent("quote amount greater than balance", trace.WithAttributes(
+	// Clip the quoteAmount by the max origin balance
+	maxBalance := m.config.GetMaxBalance(input.OriginChainID, input.OriginTokenAddr)
+	if input.OriginBalance != nil && maxBalance.Cmp(big.NewInt(0)) > 0 {
```
Comment (ChiTimesChi) [inline services/rfq/relayer/quoter/quoter.go:555]: See comment above re: default value for `maxBalance` - this brings the question how do we want to treat `maxBalance == 0` case

File: services/rfq/relayer/relconfig/getters.go (hunk 3 above)
Comment (dwasse) [reply services/rfq/relayer/relconfig/getters.go:301]: Good idea, will use `nil` for this

File: services/rfq/relayer/quoter/quoter.go (hunk 4 above)
Comment (dwasse) [reply services/rfq/relayer/quoter/quoter.go]: Current thinking is to use `nil` for 'positive inf'- this way an unspecified `max_balance` param in the config will mean no max balance value, explicitly setting it to zero should result in all quotes to this chain having 0 quoteAmount

File: services/rfq/relayer/quoter/quoter.go [hunk 5]
```diff
@@ -519,17 +550,39 @@ func (m *Manager) getOriginAmount(parentCtx context.Context, origin, dest int, a
...
+				attribute.String("max_balance", maxBalance.String()),
+				attribute.String("origin_balance", input.OriginBalance.String()),
+			))
+			quoteAmount = quotableBalance
+		}
+	}
+
+	// Finally, clip the quoteAmount by the dest balance
+	if quoteAmount.Cmp(input.DestBalance) > 0 {
+		span.AddEvent("quote amount greater than quotable balance", trace.WithAttributes(
```
Comment (ChiTimesChi) [inline services/rfq/relayer/quoter/quoter.go:577]: Should be "greater than destination balance"?

**Review by ChiTimesChi — APPROVED** [review]

**Pull request merged**